keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,551 @@
+
+For example, a Job with a `Parallelism`: 4 and `Ready` of 3. Means that 1 pod of a Job is not ready.  
+The Job API has no concept of an overall readiness condition based on a percentage of pods that are ready.  
+
+We could 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+```golang
+// Status of ReplicatedJob for StartupPolicy
+type Status string
+
+const (
+ // Ready means that the job is ready
+ Ready Status = "Ready"
+
+ // Succeeded means that the job is successful
+ Succeeded Status = "Succeeded"
+)
+
+type StartupPolicy struct {
+ // TargetReplicatedJobStatuses is a list of rules for applying StartupPolicy.
+ // +optional
+ // +listType=map
+ // +listMapKey=name
+ TargetReplicatedJobStatuses []TargetReplicatedJobStartup `json:"targetReplicatedJobStatuses,omitempty"`
+}
+
+type TargetReplicatedJobStartup struct {
+ // Name of ReplicatedJob to apply startup policy
+ Name string `json:"name"`
+
+ // The status you want the ReplicatedJob to be
+ // +kubebuilder:validation:Enum=Ready;Succeeded
+ Status Status `json:"status"`
+
+ // Number of Statuses is how many ReplicatedJob statuses to consider "started"
+ // Empty will mean all status must match the replicas.
+ NumberOfStatuses int32 `json:"numberOfStatuses,omitempty"`
+}
+```
+
+Keeping the non specified jobs halted we propose to use suspend for these jobs.
+
+So we will add a new field to the ReplicatedJobStatus.
+
+```golang
+// ReplicatedJobStatus defines the observed ReplicatedJobs Readiness.
+type ReplicatedJobStatus struct {
+  ...
+	Suspended int32  `json:"suspended"`
+}
+```
+
+### Implementation
+
+If StartupPolicy is set and suspend field is off, we will start all jobs in a suspended state.  We will not set a condition for JobSet as suspended.
+
+To track which ReplicatedJobs are suspended, we will add a `Suspended` field to the ReplicatedJobStatus.  This field will track the number of replicas that are suspended.
+
+When a JobSet is started with `StartupPolicy` specified, we will create jobs in a suspended state (ie `Job.Spec.Suspend = true`)  This avoids starting the underlying jobs.
+
+If StartupPolicy is set, then we will suspend the other jobs until the StartupPolicy is considered succesful.
+We will unsuspend only the jobs that match `TargetReplicatedJobStartup`.  Once they are considered "started" by matching on targeted status, then we resume all the rest of the jobs. 
```

Comment:
> Or is this saying that we handle only 1 startup policy at a time, in the order in which they are specified in the spec (e.g. we first see a startup policy for Job A so we unsuspend it, wait for Job A to complete, then we check the next startup policy and it's for Job B, so we unsuspend that, etc.)?

This is where I find it difficult balancing design versus implementation details.  

In the implementation, I apply StartupPolicy in sequential order.  

I would verify if JobA is suspended and then unsuspend.  I would return and trigger a reconcile.  

I loop over the startupPolicy rules and check if JobA is successful via the rules (Ready/Succeeded) and if it is, I move on to JobB and resume if its suspended.  

And repeat.  

If all the startup polices items are successful then we unsuspend everything not in the startup list (because the startup ones are running or succeeded).

Reply:
updated with this info. PTAL